debugd/internal/debugd/deploy/service.go

```diff
@@ -149,28 +159,38 @@ func (s *ServiceManager) WriteSystemdUnitFile(ctx context.Context, unit SystemdU
 	}
 
 	log.Infof("Wrote systemd unit file and performed daemon-reload")
-
 	return nil
 }
 
-// DefaultServiceUnit will write the default "bootstrapper.service" unit file.
-func DefaultServiceUnit(ctx context.Context, serviceManager *ServiceManager) error {
-	if err := serviceManager.WriteSystemdUnitFile(ctx, SystemdUnit{
-		Name:     debugd.BootstrapperSystemdUnitName,
-		Contents: debugd.BootstrapperSystemdUnitContents,
-	}); err != nil {
-		return fmt.Errorf("writing systemd unit file %q: %w", debugd.BootstrapperSystemdUnitName, err)
+// OverrideServiceUnitExecStart will override the ExecStart of a systemd unit.
+func (s *ServiceManager) OverrideServiceUnitExecStart(ctx context.Context, unitName, execStart string) error {
+	log := s.log.With(zap.String("unitFile", fmt.Sprintf("%s/%s", systemdUnitFolder, unitName)))
+	log.Infof("Overriding systemd unit file execStart")
+	if !systemdUnitNameRegexp.MatchString(unitName) {
+		return fmt.Errorf("unit name %q is invalid", unitName)
 	}
-
-	// try to start the default service if the binary exists but ignore failure.
-	// this is meant to start the bootstrapper after a reboot
-	// if a bootstrapper binary was uploaded before.
-	if ok, err := afero.Exists(serviceManager.fs, debugd.BootstrapperDeployFilename); ok && err == nil {
-		_ = serviceManager.SystemdAction(ctx, ServiceManagerRequest{
-			Unit:   debugd.BootstrapperSystemdUnitName,
-			Action: Start,
-		})
+	// validate execStart (no newlines)
+	if strings.Contains(execStart, "\n") || strings.Contains(execStart, "\r") {
+		return fmt.Errorf("execStart must not contain newlines")
+	}
+	overrideUnitContents := fmt.Sprintf("[Service]\nExecStart=\nExecStart=%s\n", execStart)
+	s.systemdUnitFilewriteLock.Lock()
+	defer s.systemdUnitFilewriteLock.Unlock()
+	path := filepath.Join(systemdUnitFolder, unitName+".service.d", "override.conf")
+	if err := s.fs.MkdirAll(filepath.Dir(path), os.ModePerm); err != nil {
+		return fmt.Errorf("creating systemd unit file override directory %q: %w", filepath.Dir(path), err)
+	}
+	if err := afero.WriteFile(s.fs, path, []byte(overrideUnitContents), 0o644); err != nil {
+		return fmt.Errorf("writing systemd unit override file %q: %w", unitName, err)
+	}
+	if err := s.SystemdAction(ctx, ServiceManagerRequest{Unit: unitName, Action: Reload}); err != nil {
+		log.Warnf("Failed to perform systemd daemon-reload: %v", err)
+	}
+	if err := s.SystemdAction(ctx, ServiceManagerRequest{Unit: unitName + ".service", Action: Restart}); err != nil {
+		log.Warnf("Failed to perform unit restart: %v", err)
+		return fmt.Errorf("performing systemd unit restart: %w", err)
 	}
```

Comment:
Is there a use-case to allow restarting when reloading fails? Otherwise I'd also return with an error for the latter.

Reply:
Reloading systemd is inherently flaky in real life. I saw different errors at this point that are not actual errors and there is also no signal from the API to see if the reload was actually successful.
This is why I decided to log this info but just continue with the restart in any case.

Reply:
Ok, I assume reloading until truly no error returns is a bit much in this case? Then I am fine with the current solution, since it is for internal use only.

Reply:
I'll still add a comment to explain this behavior